# Electronic Store Sales : Inital Data Cleaning
Step 2 - Prepare data for analysis
- Project Goals:Look at the sales data of an electronic store. See if there is any way to increase sales.

- Goals of this Notebook: clean data

- Goals of next notebooks: run analysis, create visualisations, conclusions

Cleaning to do 
1. Clean 545 null values
2. Seperate the date from the time in the order data column
3. Change data types on quantity ordered, price, order date
4. Seperate the address into cities and states

Potential Analysis
1. Look for correlations with quant ordered and date/city/state/price
2. Find the cities with the top sales
3. Find the time for most sales
4. Filter sales by order ID to find which products are often sold together

### 1. Import Libraries and Data

In [1]:
import pandas as pd
import os

In [2]:
df = pd.read_csv('all_data.csv')
df.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   

       Order Date                      Purchase Address  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001  
1             NaN                                   NaN  
2  04/07/19 22:30     682 Chestnut St, Boston, MA 02215  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  
4  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001

#### Get rid of the NaN values
Create a df of all the rows with NaN values to examine what is missing from the entries. 

In [12]:
df_nan = df[df.isna().any(axis=1)]
df_nan.head()

Order ID Product Quantity Ordered Price Each Order Date Purchase Address
1         NaN     NaN              NaN        NaN        NaN              NaN
356       NaN     NaN              NaN        NaN        NaN              NaN
735       NaN     NaN              NaN        NaN        NaN              NaN
1433      NaN     NaN              NaN        NaN        NaN              NaN
1553      NaN     NaN              NaN        NaN        NaN              NaN

We see that these rows have no data, not just a missing column, so lets drop them from our all_data df using dropna syntax. 

In [13]:
df = df.dropna(how = 'all')  # how = 'all' means that we are droping rows where nan is in all the columns of that row.
df.isnull().sum()

Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64

### Seperate the date, Add a Month Column

In [54]:
df['Month'] = df['Order Date'].apply(lambda x: x[:2]) 
df.sample(5)

<ipython-input-54-5bf718f19ddd>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Month'] = df['Order Date'].apply(lambda x: x[:2])


Order ID                 Product Quantity Ordered Price Each  \
167310   271316        Wired Headphones                1      11.99   
67648    141333        27in FHD Monitor                1     149.99   
58029    152904           Flatscreen TV                1        300   
3529     179942                LG Dryer                1      600.0   
2336     178791  AAA Batteries (4-pack)                1       2.99   

            Order Date                        Purchase Address Month  
167310  10/11/19 22:37   384 Dogwood St, Los Angeles, CA 90001    10  
67648   01/01/19 13:58      554 10th St, Los Angeles, CA 90001    01  
58029   02/15/19 21:58         480 Maple St, Atlanta, GA 30301    02  
3529    04/16/19 14:24  209 Meadow St, New York City, NY 10001    04  
2336    04/27/19 12:56    28 Hickory St, Los Angeles, CA 90001    04

The warning is a false warning? not sure about this can't seem to get rid of it

### Convert Month Column to int

In [60]:
# getting an error when trying to convert astpye int
# invalid literal for int() with base 10: 'Or'
df['Month'].unique()

array(['04', '05', 'Or', '08', '09', '12', '01', '02', '03', '07', '06',
       '11', '10'], dtype=object)

In [62]:
or_df = df[df['Order Date'].str[0:2] == 'Or']
or_df

Order ID  Product  Quantity Ordered  Price Each  Order Date  \
519     Order ID  Product  Quantity Ordered  Price Each  Order Date   
1149    Order ID  Product  Quantity Ordered  Price Each  Order Date   
1155    Order ID  Product  Quantity Ordered  Price Each  Order Date   
2878    Order ID  Product  Quantity Ordered  Price Each  Order Date   
2893    Order ID  Product  Quantity Ordered  Price Each  Order Date   
...          ...      ...               ...         ...         ...   
185164  Order ID  Product  Quantity Ordered  Price Each  Order Date   
185551  Order ID  Product  Quantity Ordered  Price Each  Order Date   
186563  Order ID  Product  Quantity Ordered  Price Each  Order Date   
186632  Order ID  Product  Quantity Ordered  Price Each  Order Date   
186738  Order ID  Product  Quantity Ordered  Price Each  Order Date   

        Purchase Address Month  
519     Purchase Address    Or  
1149    Purchase Address    Or  
1155    Purchase Address    Or  
2878    Purchase Address    Or  
2893    Purchase Address    Or  
...                  ...   ...  
185164  Purchase Address    Or  
185551  Purchase Address    Or  
186563  Purchase Address    Or  
186632  Purchase Address    Or  
186738  Purchase Address    Or  

[355 rows x 7 columns]

There are 355 invalid rows we need to drop. 

In [64]:
df = df[df['Order Date'].str[0:2] != 'Or']
df['Month'] = df['Month'].astype('int')
df.sample(5)

Order ID                     Product Quantity Ordered Price Each  \
81270    226732        USB-C Charging Cable                1      11.95   
135977   208839      AAA Batteries (4-pack)                2       2.99   
966      177477      AAA Batteries (4-pack)                1       2.99   
42309    307025  Bose SoundSport Headphones                1      99.99   
142718   284143    Lightning Charging Cable                2      14.95   

            Order Date                       Purchase Address  Month  
81270   07/25/19 16:36      520 5th St, Los Angeles, CA 90001      7  
135977  05/19/19 13:36        199 Church St, Austin, TX 73301      5  
966     04/03/19 19:49    339 5th St, New York City, NY 10001      4  
42309   12/23/19 14:38    394 4th St, San Francisco, CA 94016     12  
142718  11/05/19 07:31  554 River St, San Francisco, CA 94016     11

### Covert other columns to correct type

In [70]:
df['Quantity Ordered'] = df['Quantity Ordered'].astype(int)
df['Price Each'] = df['Price Each'].astype(float)
print(df.dtypes)

Order ID                    object
Product                     object
Quantity Ordered             int32
Price Each                 float64
Order Date          datetime64[ns]
Purchase Address            object
Month                        int32
City                        object
Hour                         int64
Minute                       int64
dtype: object


### Create a City, State column

In [65]:
df['City'] = df['Purchase Address'].apply(lambda x: x.split(',')[1] + '' + x.split(',')[2][:3])
df.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   
5   176561            Wired Headphones                1      11.99   

       Order Date                      Purchase Address  Month  \
0  04/19/19 08:46          917 1st St, Dallas, TX 75001      4   
2  04/07/19 22:30     682 Chestnut St, Boston, MA 02215      4   
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001      4   
4  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001      4   
5  04/30/19 09:27     333 8th St, Los Angeles, CA 90001      4   

              City  
0        Dallas TX  
2        Boston MA  
3   Los Angeles CA  
4   Los Angeles CA  
5   Los Angeles CA

### Create Hour and Minute Purchased Columns

In [68]:
# change format of Order Date column to make extraction easier
df['Order Date'] = pd.to_datetime(df['Order Date'])

# add the Hour and Minute columns.
df['Hour'] = df['Order Date'].dt.hour
df['Minute'] = df['Order Date'].dt.minute

df.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   
5   176561            Wired Headphones                1      11.99   

           Order Date                      Purchase Address  Month  \
0 2019-04-19 08:46:00          917 1st St, Dallas, TX 75001      4   
2 2019-04-07 22:30:00     682 Chestnut St, Boston, MA 02215      4   
3 2019-04-12 14:38:00  669 Spruce St, Los Angeles, CA 90001      4   
4 2019-04-12 14:38:00  669 Spruce St, Los Angeles, CA 90001      4   
5 2019-04-30 09:27:00     333 8th St, Los Angeles, CA 90001      4   

              City  Hour  Minute  
0        Dallas TX     8      46  
2        Boston MA    22      30  
3   Los Angeles CA    14      38  
4   Los Angeles CA    14      38  
5   Los Angeles CA     9      27

### Save cleaned df

In [71]:
df.to_csv("all_data_cleaned.csv", index=False)